File: backend-csharp/CalendarDay.Infrastructure/Services/AuditLogsService.cs
using CalendarDay.Application.Abstractions;
using CalendarDay.Application.Contracts;
using CalendarDay.Application.Contracts.AuditLogs;
using CalendarDay.Domain.Entities;
using CalendarDay.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace CalendarDay.Infrastructure.Services;

public class AuditLogsService(CalendarDayDbContext db) : IAuditLogsService
{
    public async Task<PagedResult<AuditLogDto>> GetAsync(AuditLogQuery query, CancellationToken ct)
    {
        var page = Math.Max(1, query.Page);
        var pageSize = Math.Clamp(query.PageSize, 1, 200);

        IQueryable<AuditLog> q = db.AuditLogs.AsNoTracking();

        if (!string.IsNullOrWhiteSpace(query.Search))
        {
            var search = query.Search.Trim().ToLowerInvariant();
            q = q.Where(x =>
                (x.Username ?? string.Empty).ToLower().Contains(search) ||
                x.Action.ToLower().Contains(search) ||
                (x.Details ?? string.Empty).ToLower().Contains(search) ||
                x.Endpoint.ToLower().Contains(search) ||
                x.Method.ToLower().Contains(search));
        }

        if (!string.IsNullOrWhiteSpace(query.User))
        {
            var user = query.User.Trim().ToLowerInvariant();
            q = q.Where(x => (x.Username ?? string.Empty).ToLower().Contains(user));
        }
        if (!string.IsNullOrWhiteSpace(query.Action))
        {
            var action = query.Action.Trim().ToLowerInvariant();
            q = q.Where(x => x.Action.ToLower().Contains(action));
        }
        if (!string.IsNullOrWhiteSpace(query.Endpoint))
        {
            var endpoint = query.Endpoint.Trim().ToLowerInvariant();
            q = q.Where(x => x.Endpoint.ToLower().Contains(endpoint));
        }
        if (query.StatusCode.HasValue)
        {
            q = q.Where(x => x.StatusCode == query.StatusCode.Value);
        }
        if (query.FromUtc.HasValue)
        {
            q = q.Where(x => x.CreatedAtUtc >= query.FromUtc.Value);
        }
        if (query.ToUtc.HasValue)
        {
            q = q.Where(x => x.CreatedAtUtc <= query.ToUtc.Value);
        }

        var sortBy = query.SortBy.Trim().ToLowerInvariant();
        var sortAsc = query.SortDir.Trim().Equals("asc", StringComparison.OrdinalIgnoreCase);
        q = (sortBy, sortAsc) switch
        {
            ("status", true) => q.OrderBy(x => x.StatusCode).ThenBy(x => x.CreatedAtUtc),
            ("status", false) => q.OrderByDescending(x => x.StatusCode).ThenByDescending(x => x.CreatedAtUtc),
            ("user", true) => q.OrderBy(x => x.Username).ThenBy(x => x.CreatedAtUtc),
            ("user", false) => q.OrderByDescending(x => x.Username).ThenByDescending(x => x.CreatedAtUtc),
            ("action", true) => q.OrderBy(x => x.Action).ThenBy(x => x.CreatedAtUtc),
            ("action", false) => q.OrderByDescending(x => x.Action).ThenByDescending(x => x.CreatedAtUtc),
            _ when sortAsc => q.OrderBy(x => x.CreatedAtUtc),
            _ => q.OrderByDescending(x => x.CreatedAtUtc),
        };

        var totalCount = await q.CountAsync(ct);
        var items = await q
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(x => new AuditLogDto(
                x.Id,
                x.Username,
                x.Action,
                x.Details,
                x.Endpoint,
                x.Method,
                x.StatusCode,
                x.IpAddress,
                x.CreatedAtUtc))
            .ToListAsync(ct);

        return new PagedResult<AuditLogDto>(items, page, pageSize, totalCount);
    }

    public async Task LogAsync(string? username, string action, string? details, string endpoint, string method, int statusCode, string? ipAddress, CancellationToken ct)
    {
        var log = new AuditLog
        {
            Id = Guid.NewGuid(),
            Username = string.IsNullOrWhiteSpace(username) ? null : username.Trim().ToLowerInvariant(),
            Action = action.Trim(),
            Details = string.IsNullOrWhiteSpace(details) ? null : details.Trim(),
            Endpoint = endpoint.Trim(),
            Method = method.Trim().ToUpperInvariant(),
            StatusCode = statusCode,
            IpAddress = ipAddress,
            CreatedAtUtc = DateTime.UtcNow,
        };
        db.AuditLogs.Add(log);
        await db.SaveChangesAsync(ct);
    }
}

File: src/features/auditLogs/services/auditLogsService.ts
import { apiRequest } from '../../../shared/services/apiClient';

export type AuditLogItem = {
  id: string;
  username: string | null;
  action: string;
  details: string | null;
  endpoint: string;
  method: string;
  statusCode: number;
  ipAddress: string | null;
  createdAtUtc: string;
};

export type AuditLogsQuery = {
  page: number;
  pageSize: number;
  search?: string;
  user?: string;
  action?: string;
  endpoint?: string;
  fromUtc?: string;
  toUtc?: string;
  statusCode?: string;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
};

export type PagedAuditLogs = {
  items: AuditLogItem[];
  page: number;
  pageSize: number;
  totalCount: number;
};

export async function fetchAuditLogs(query: AuditLogsQuery, signal?: AbortSignal): Promise<PagedAuditLogs> {
  const params = new URLSearchParams();
  params.set('page', String(query.page));
  params.set('pageSize', String(query.pageSize));
  if (query.search?.trim()) params.set('search', query.search.trim());
  if (query.user?.trim()) params.set('user', query.user.trim());
  if (query.action?.trim()) params.set('action', query.action.trim());
  if (query.endpoint?.trim()) params.set('endpoint', query.endpoint.trim());
  if (query.fromUtc?.trim()) params.set('fromUtc', query.fromUtc.trim());
  if (query.toUtc?.trim()) params.set('toUtc', query.toUtc.trim());
  if (query.statusCode?.trim()) params.set('statusCode', query.statusCode.trim());
  if (query.sortBy?.trim()) params.set('sortBy', query.sortBy.trim());
  if (query.sortDir?.trim()) params.set('sortDir', query.sortDir.trim());

  return apiRequest<PagedAuditLogs>(`/audit-logs?${params.toString()}`, { signal });
}
